.github/workflows/pre-commit.yaml

```diff
@@ -0,0 +1,45 @@
+name: pre-commit
+
+on:
+  pull_request:
+  push:
+    branches:
+      - main
+      # For development, and debugging of the workflow.
+      - force_ci/pre-commit/**
+
+permissions:
+  contents: read
+  pull-requests: write
+
+jobs:
+  pre-commit:
+    runs-on: ubuntu-latest
+    steps:
+      - id: checkout
+        uses: actions/checkout@v4
+
+      - id: setup-python
+        uses: actions/setup-python@v5
+        with:
+          python-version: 3.11
+
+      - name: Install shellcheck, shfmt
+        run: sudo apt-get install -y shellcheck shfmt
```

Comment:
One final note: by getting these from whatever `ubuntu-latest` has, team members might be painted into a corner when trying to run CI locally (might need a VM or a container, or `dpkg -i`).

This has no easy solution, though, and is good enough for me for now.

(I'd personall consider installing from Linuxbrew, which always tracks the most recent upstream.)

Reply:
Might it make sense to target debian stable or testing instead? So at least it could be identical to dev environments?